packages/web/src/store/application/ui/buy-audio/sagas.ts

```diff
@@ -0,0 +1,421 @@
+            return 'Insufficient funds'
+          }
+          return `Error Code: ${errorCode}`
+        }
+      }
+    }
+  )
+  if (result.error) {
+    throw new Error(`${name} transaction failed: ${result.error}`)
+  }
+  console.debug(`Exchange: ${name} transaction... success txid: ${result.res}`)
+  return result
+}
+
+/**
+ * Executes a Jupiter Swap given the RouteInfo, account, and the transactionHandler
+ */
+function* doSwap({
+  route,
+  account,
+  transactionHandler
+}: {
+  route: RouteInfo
+  account: Keypair
+  transactionHandler: TransactionHandler
+}) {
+  const jup = yield* call(initJupiterIfNecessary)
+  const {
+    transactions: { setupTransaction, swapTransaction, cleanupTransaction }
+  } = yield* call([jup, jup.exchange], {
+    routeInfo: route,
+    userPublicKey: account.publicKey
+  })
+  if (setupTransaction) {
+    yield* call(
+      sendTransaction,
+      'Setup',
+      setupTransaction,
+      account,
+      transactionHandler
+    )
+  }
+  // Wrap this in try/finally to ensure cleanup transaction runs, if applicable
+  try {
+    yield* call(
+      sendTransaction,
+      'Swap',
+      swapTransaction,
+      account,
+      transactionHandler
+    )
+  } finally {
+    if (cleanupTransaction) {
+      yield* call(
+        sendTransaction,
+        'Cleanup',
+        cleanupTransaction,
+        account,
+        transactionHandler
+      )
+    }
+  }
+}
+
+/**
+ * Swaps the input token (eg. SOL) to the output token (AUDIO) and, if necessary,
+ * transfers resulting AUDIO into the user's userbank from their root Solana account
+ *
+ * Currently only supports SOL => AUDIO
+ *
+ * @returns the amount of AUDIO transferred to the userbank
+ */
+function* doExchange({
+  inputAmount,
+  inputTokenSymbol,
+  outputTokenSymbol,
+  slippage
+}: ReturnType<typeof exchange>['payload']) {
+  const connection: Connection = yield* call(getSolanaConnection)
+  const rootAccount: Keypair = yield* call(getRootSolanaAccount)
+
+  // Only supports SOL => AUDIO for now
+  if (inputTokenSymbol !== 'SOL' || outputTokenSymbol !== 'AUDIO') {
+    throw new Error(
+      `Exchange not supported for ${inputTokenSymbol} => ${outputTokenSymbol}`
+    )
+  }
+
+  // Get a fresh quote - reduces chances of failing due to slippage,
+  // but increases chance of different price than originally quoted (outputted $AUDIO can vary more).
+  // It's also required since we can't put RouteInfo in redux (it isn't a plain object), so the saga return value is used.
+  const exchangeQuote = yield* call(doQuote, {
+    inputAmount,
+    inputTokenSymbol,
+    outputTokenSymbol,
+    forceFetch: true,
+    slippage
+  })
+  const transactionHandler = new TransactionHandler({
+    connection,
+    useRelay: false,
+    feePayerKeypairs: [rootAccount],
+    skipPreflight: true
+  })
+  yield* call(doSwap, {
+    route: exchangeQuote.route!,
+    account: rootAccount,
+    transactionHandler
+  })
+  const account = yield* select(getAccountUser)
```

Comment:
maybe do this check in the first few lines of the function so we don't have to init TransactionHandler or call doSwap